cluster-autoscaler/utils/gpu/gpu_test.go

```diff
@@ -31,54 +34,191 @@ const (
+			wantLabel:    "generic",
 		},
-	}
-	assert.True(t, NodeHasGpu(GPULabel, nodeGpuUnready))
-
-	nodeNoGpu := &apiv1.Node{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:   "nodeNoGpu",
-			Labels: map[string]string{},
+		{
+			desc: "invalid gpu type",
+			config: &cloudprovider.GpuConfig{
+				Label:        "label/gpu-type",
+				Type:         "no-such-gpu",
+				ResourceName: util.ResourceNvidiaGPU,
+			},
+			gpus: map[string]struct{}{
+				"gpu-type-value": {},
+			},
+			node:         newNode("label/gpu-type", "no-such-gpu", resource.NewQuantity(10, resource.DecimalSI)),
+			templateNode: newNode("label/gpu-type", "no-such-gpu", resource.NewQuantity(10, resource.DecimalSI)),
+			wantResource: util.ResourceNvidiaGPU,
+			wantLabel:    "not-listed",
+		},
+		{
+			desc: "happy scenario",
+			config: &cloudprovider.GpuConfig{
+				Label:        "label/gpu-type",
+				Type:         "gpu-type-value",
+				ResourceName: util.ResourceNvidiaGPU,
+			},
+			gpus: map[string]struct{}{
+				"gpu-type-value": {},
+			},
+			node:         newNode("label/gpu-type", "gpu-type-value", resource.NewQuantity(10, resource.DecimalSI)),
+			templateNode: newNode("label/gpu-type", "gpu-type-value", resource.NewQuantity(10, resource.DecimalSI)),
+			wantResource: util.ResourceNvidiaGPU,
+			wantLabel:    "gpu-type-value",
+		},
+		{
+			desc: "missing gpu",
+			config: &cloudprovider.GpuConfig{
+				Label:        "label/gpu-type",
+				Type:         "gpu-type-value",
+				ResourceName: util.ResourceNvidiaGPU,
+			},
+			gpus: map[string]struct{}{
+				"gpu-type-value": {},
+			},
+			node:         newNode("label/gpu-type", "gpu-type-value", nil),
+			templateNode: newNode("label/gpu-type", "gpu-type-value", resource.NewQuantity(10, resource.DecimalSI)),
+			wantResource: util.ResourceNvidiaGPU,
+			wantLabel:    "missing-gpu",
 		},
-		Status: apiv1.NodeStatus{
-			Capacity:    apiv1.ResourceList{},
-			Allocatable: apiv1.ResourceList{},
+		{
+			desc: "no capacity",
+			config: &cloudprovider.GpuConfig{
+				Label:        "label/gpu-type",
+				Type:         "gpu-type-value",
+				ResourceName: util.ResourceNvidiaGPU,
+			},
+			gpus: map[string]struct{}{
+				"gpu-type-value": {},
+			},
+			node:         newNode("label/gpu-type", "gpu-type-value", nil),
+			templateNode: newNode("label/gpu-type", "gpu-type-value", nil),
+			wantResource: util.ResourceNvidiaGPU,
+			wantLabel:    "unexpected-label",
+		},
+		{
+			desc: "no capacity and node group",
+			config: &cloudprovider.GpuConfig{
+				Label:        "label/gpu-type",
+				Type:         "gpu-type-value",
+				ResourceName: util.ResourceNvidiaGPU,
+			},
+			gpus: map[string]struct{}{
+				"gpu-type-value": {},
+			},
+			node:         newNode("label/gpu-type", "gpu-type-value", nil),
+			wantResource: util.ResourceNvidiaGPU,
+			wantLabel:    "unexpected-label",
 		},
 	}
-	assert.False(t, NodeHasGpu(GPULabel, nodeNoGpu))
-}
+	for _, tc := range tests {
+		func(tc testCase) {
```

Comment:
Why is this additional func wrapper here? I don't really see why this is needed.

Reply:
Replaced with a simpler pattern of tc := tc